code-studio/build-commands.js

```diff
@@ -41,141 +50,103 @@ exports.bundle = function (config) {
-  // --compress performs lots of optimizations including removal of dead code.
-  // warnings=false disables compressor warnings (which don't indicate actual problems, for us)
-  // --mangle replaces variable names with short representations
-  var uglifyCommand = 'uglifyjs --compress warnings=false --mangle';
+  // babelify tranforms jsx files for us
+  bundler.transform('babelify', {
+    compact: false
+  });
 
-  var factorStep = '';
+  // factor-bundle splits the bundle back up into parts, but leaves common
+  // modules in the main browserify stream.
   if (shouldFactor) {
-    // We use command substitution to build an output filename based on
-    // the input filename, which factor-bundle automatically binds to $FILE.
-    var factoredOutputFilename = buildPath + '`';
+    bundler.plugin('factor-bundle', {
+      outputs: filenames.map(function (file) {
+        return outPath(srcPath + file);
+      })
+    });
+  }
 
-    // The output file ends up at the same path relative to the build directory
-    // as the input file is relative to the src directory:
-    // @see `man realpath`
-    factoredOutputFilename += 'basename $FILE';
+  return new Promise(function (resolve) {
+    var isResolved = false;
+    var resolveOnce = function (result) {
+      if (isResolved) { return; }
+      resolve(result);
+      isResolved = true;
+    };
 
-    // And we swap out the input file extension (.js or .jsx) for an output
-    // file extension (.js or .min.js) depending on our configuration.
-    // @see `man sed`
-    factoredOutputFilename += ' | sed -r "s/\.jsx?$/' + extension + '/i"';
+    // TODO: release/dist settings
 
-    // Finally, we close the command substitution
-    factoredOutputFilename += '`';
+    // Define output step
+    var runBundle = function () {
+      var bundlingAttemptError;
 
-    // Use bracket syntax to invoke the factor-bundle plugin with one argument:
-    // A command that accepts each factored output file as it's processed.
-    factorStep = "-p [ factor-bundle -o '" +
-        (shouldMinify ? uglifyCommand : '') +
-        ' > ' + factoredOutputFilename + "' ]";
-  }
-  var commonOutput = (shouldMinify ? '| ' + uglifyCommand + ' ' : '') +
-    '-o ' + buildPath + path.basename(commonFile, '.js') + extension;
+      // We attach events to our filesystem output stream, because it's the
+      // best indicator of when the bundle is actually done building.
+      var outStream = fs.createWriteStream(outPath(srcPath + commonFile))
+          .on('error', function (err) {
+            logBoldRedError(err);
+            resolveOnce(err);
+          })
+          .on('finish', function () {
+            if (bundlingAttemptError) {
+              resolveOnce(bundlingAttemptError);
+            } else {
+              console.log(timeStamp() + ' Built ' +
+                  targets.map(function (target) {
+                    return path.join(buildPath, path.relative(buildPath, target));
```

Comment:
We end up getting awfully nested in some of these places. I'm not sure I have any great solutions on how to avoid that though (other than always indenting 2 instead of sometimes indenting 4)